File: usvm-jvm-concrete/agent/src/main/java/org/usvm/jvm/concrete/agent/JcBytecodeGetter.java
package org.usvm.jvm.concrete.agent;

import org.usvm.jvm.concrete.JcConcreteClassLoader;

import java.lang.instrument.ClassFileTransformer;
import java.lang.instrument.IllegalClassFormatException;
import java.security.ProtectionDomain;

public class JcBytecodeGetter implements ClassFileTransformer {
    @Override
    public byte[] transform(ClassLoader loader,
                            String className,
                            Class<?> classBeingRedefined,
                            ProtectionDomain protectionDomain,
                            byte[] classfileBuffer) throws IllegalClassFormatException {
        if (className == null || !(loader instanceof JcConcreteClassLoader cl)) {
            return classfileBuffer;
        }

        cl.addTypeBytes(className, classfileBuffer);

        return classfileBuffer;
    }
}

File: usvm-jvm-concrete/src/main/kotlin/features/JcBytecodeGetter.kt
package features

import machine.JcConcreteMemoryClassLoader
import utils.typeIsRuntimeGenerated
import java.lang.instrument.ClassFileTransformer
import java.security.ProtectionDomain

class JcBytecodeGetter : ClassFileTransformer {

    override fun transform(
        loader: ClassLoader?,
        className: String?,
        classBeingRedefined: Class<*>?,
        protectionDomain: ProtectionDomain?,
        classfileBuffer: ByteArray
    ): ByteArray {
        if (className == null || !className.typeIsRuntimeGenerated || loader !is JcConcreteMemoryClassLoader)
            return classfileBuffer

        JcGeneratedTypesFeature.addGeneratedTypeBytes(className, classfileBuffer)

        return classfileBuffer
    }
}
